### DISCUSSION TITLE
Add Palette color keyframe data to Function Editor

### DISCUSSION THREAD

**Discussion by artisteacher** [opentoonz/opentoonz#6708] [Ideas]
**Describe the solution you'd like**
It would be helpful for color keyframes to be visible and adjustable through the Function Editor. At minimum, there should be the capability to reposition the color keyframes. It would also be great to be able to adjust the interpolation of the color keys.
 
Since the function editor can be crowded, I'd make visibility of the color keyframe data context sensitive - ie. if there are no color keyframes, no color category would display in the function editor. Once a color keyframe was added, then a color keyframe column & etc. would become visible.

**Comment by RodneyBaker** [comment]
I'd like to see this too and will add this note to describe some of the difficulty involved for the developer who takes this on as well as the potential for other enhancements once the obstacles involved are removed.

Current Situation
Unlike regular keyframes accessed via Function Editor which have their data stored in Scene files, Color Style keyframe data is stored in external file; palette files.  Taking a look at this should present not only the major obstacles involved in bringing the key data into the Function Editor but highlight the significant potential behind bridging the gap between the two keyframe storage methodologies.  Namely, that allowing color style keyframes to be viewed and edited in the function editor would significantly enhance workflow when working with color styles but would also lead to a method where standard keyframes (as used in timeline/xsheet) could be stored in external files (basically a palette file but specifically designated for storing keyframe data).  

This would be a significant undertaking but the basic functionality desired... that of allowing color style keys to be exposed in the  Function Editor... is present in the form of such interpolation methods as 'File Interpolation' where external file data can be accessed in the Function Editor.

The road ahead
In order to implement this proposed phase of Function Editor enhancement the Function Editor will need to gain some new capabilities.  It may be ideal to look at this from the broader view of how to enhance current access the Function Editor already possesses in reading data from external files in order to have that data read and updated in real time (i.e. link the data from the external file) without having to constantly and manually apply the change via the press of a button.  Instead, when the data in the external palette file changes the keyframe data of the color styles will need to automatically refresh the data linked into the Scene files.  In short, the Scene files will likely need to have an active link to the external palette file where the keyframe data resides.  From there it is a logical evolution to allow keyframe data to be stored in similar files not exclusive to palette file data but instead holding keyframe data for any object.  These external 'action files' for lack of better term could then store data that can drive changes to any object regardless of whether they are color styles, columns/levels, fx, etc.   They could then stored for use and reuse and swapped out easily as needed after being used wherever keyframe data can be manipulated in Opentoonz.

**Comment by adamearle** [comment]
has the turtle done this?

**Comment by RodneyBaker** [comment]
No, Jeremy hasn't done this.

A difficulty in realizing color keyframes in the Function Editor is that they keyframes are stored in palette files and not in scene files.  This is both a strength and a weakness; a strength, because it represents how external keyframes can be stored and a weakness, because a separate process has to be created in order to work with the data.

From what I can tell in order to make this work both approaches (color keyframes stored in palette files) and motion keyframes (stored in scene files) have to move closer to each other and share code/processes.  A first step might be to begin with the Function Editor and have it be able to export keyframes in a compatible way as the color keyframes do with external palettes.  The Function Editor can do this (in part) by exporting .curve files and data dumps but they aren't quite the same as the palette format.

I suppose what I'm suggesting is that we need a means to translate from palette files to .curve (or data) files from the Function Editor and this basically suggests creating a new file format.

**Comment by gab3d** [comment]
Palette Styles keys should internally be a collection of several standard keys, one for each stored Style 1property at that frame (color channels, alpha, other settings..).

Esentially they cannot be easily represented in the Function Editor via the Graph Editor, but they could be handled in a simple form through the Spreadsheet.

I don't think the fact that those keys are stored in a different file (and not in the scene file) should be a problem, provided you can access the data from the Function Editor. 🤷🏻‍♂️

**Comment by RodneyBaker** [comment]
I would really love to see color keyframes exposed in the function editor but there are so many differences between color keyframes and channel keyframes... it's hard to know where to start.
Here's a quick look at the keyframes for each.
The only thing they seem to have in common is that they are included in xml formatted files:

*![image](https://user-images.githubusercontent.com/2576432/90898190-e9c23680-e38b-11ea-9bbf-e948173b5df2.png)

There are a number of ways we could move forward.
How we move forward will dictate the feature set that can accompany the keyframing process.

Perhaps of interest would be some more detailed information regarding both types of keyframes.
- On the left where we have stardard motion/channel keyframes values are stored in a single channel (in the example above it is the x (east/west) channel).
- One thing palette keyframes have over motion keyframes is that one line records multiple 'channels' (i.e. red, green, blue, alpha, etc.).  
- A plus for motion keyframes is the different interpolations (the < L >tag indicating 'linear interpolation').
- With color keyframes the only interpolation (currently) appears to be linear.
- Motion keyframes appear to be decimal/float values while color keyframes are integer.
- For motion keyframes the frame number precedes the key's value (on the same line) while for color keyframes it is part of the xml tag (e.g. <keyframe "frame = 1">.

**Comment by RodneyBaker** [comment]
@gab3d said:

> Esentially they cannot be easily represented in the Function Editor via the Graph Editor, but they could be handled in a simple form through the Spreadsheet.

This is where the Function Editor/Curve Editor would likely gain from having some of the palette keyframe capability.
Specifically, because a keyframe of a color style is a collection of channels whereas a keyframe for motion is limited to storing values in just one channel.

This suggests to me that on a hierarchical level a standard (motion) keyframe is the smaller unit of the two.
The color style approach being a set or collection of several keys per frame whereas a motion keyframe is just the value at that frame for the single channel.

This makes sense to me because 'palettes' are collections of color swatches which in turn are collections of values for various channels.  The equivalent of a 'palette' file could therefore be shoehorned to collect motion keyframes into 'action files' that collect multiple channels based on what level... or more specifically what column their transform channels belong to.

**Comment by artisteacher** [comment]
To start with, I'd be happy just to gain basic spreadsheet functionality. Right now the only way to change a color keyframe position is to remove the key, change frames, and add the key in again. It would be much easier to drag or cut/paste a color keyframe to the new frame in the Function Editor. Adding interpolation options & capability to adjust with the Curve Editor should be secondary, but would be a valuable addition.

Assuming the Function Editor gained the capability to read/write palette keyframe data, how should the keyframe data display in the spreadsheet? It might be worth using hex#, since it takes up less space than rgba.

**Comment by gab3d** [comment]
@artisteacher 
Keyframe data would not be directly accessible in the Function Editor, just the timing of the keys. 
To see and modify the actual values from the selected palette key I assume you should also have a Style Editor panel opened.

Probably, interpolation of values could be handled in Function Editor, but it would have to be the same for the whole set of properties contained in the palette keys. I suppose currently they have a Linear interpolation.
